primitives/src/sync.rs

```diff
@@ -0,0 +1,149 @@
+use core::ops::{Deref, DerefMut};
+use std::{
+    sync::atomic::{AtomicU32, Ordering},
+    time::{Duration, Instant},
+};
+
+const TIME_SPINNING_MICRO_SECONDS: u128 = 500;
+const SLEEP_TIME: Duration = Duration::from_micros(100);
+
+pub struct MutexMutGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct MutexGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct Mutex<T: ?Sized> {
+    counter: AtomicU32,
+    interior: T,
+}
+
+impl<T: ?Sized> std::fmt::Debug for Mutex<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "iroha_mutex<{}>", std::any::type_name::<T>())
+    }
+}
+
+impl<T> Mutex<T> {
+    pub fn new(interior: T) -> Self {
+        Self {
+            counter: AtomicU32::new(0),
+            interior,
+        }
+    }
+
+    pub fn lock_mut(&self) -> MutexMutGuard<T> {
+        let instant_entered = Instant::now();
+        loop {
+            match AtomicU32::compare_exchange_weak(
+                &self.counter,
+                0,
+                1,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
```

Comment:
CAREFUL! These are not the right orderings!

Reply:
I think that the success case is ok, sequential consistency is not required in the current implementation. Not sure about the failure case

Reply:
I take it back. Success should be `Acquire` and failure can be `Relaxed` ordering